src/librustc/hir/map/hir_id_validator.rs

```diff
@@ -147,33 +147,31 @@ impl<'a, 'hir: 'a> intravisit::Visitor<'hir> for HirIdValidator<'a, 'hir> {
         intravisit::NestedVisitorMap::OnlyBodies(self.hir_map)
     }
 
-    fn visit_id(&mut self, node_id: NodeId) {
+    fn visit_id(&mut self, hir_id: HirId) {
         let owner = self.owner_def_index.expect("no owner_def_index");
-        let stable_id = self.hir_map.definitions().node_to_hir_id[node_id];
 
-        if stable_id == hir::DUMMY_HIR_ID {
-            self.error(|| format!("HirIdValidator: No HirId assigned for NodeId {}: {:?}",
-                                     node_id,
-                                     self.hir_map.node_to_string(node_id)));
+        if hir_id == hir::DUMMY_HIR_ID {
+            self.error(|| format!("HirIdValidator: HirId {:?} is invalid",
+                                  self.hir_map.hir_to_string(hir_id)));
             return;
         }
 
-        if owner != stable_id.owner {
+        if owner != hir_id.owner {
             self.error(|| format!(
                 "HirIdValidator: The recorded owner of {} is {} instead of {}",
-                self.hir_map.node_to_string(node_id),
-                self.hir_map.def_path(DefId::local(stable_id.owner)).to_string_no_crate(),
+                self.hir_map.hir_to_string(hir_id),
+                self.hir_map.def_path(DefId::local(hir_id.owner)).to_string_no_crate(),
                 self.hir_map.def_path(DefId::local(owner)).to_string_no_crate()));
         }
 
-        if let Some(prev) = self.hir_ids_seen.insert(stable_id.local_id, node_id) {
-            if prev != node_id {
+        if let Some(prev) = self.hir_ids_seen.insert(hir_id.local_id, hir_id) {
+            if prev != hir_id {
                 self.error(|| format!(
                     "HirIdValidator: Same HirId {}/{} assigned for nodes {} and {}",
-                    self.hir_map.def_path(DefId::local(stable_id.owner)).to_string_no_crate(),
-                    stable_id.local_id.as_usize(),
-                    self.hir_map.node_to_string(prev),
-                    self.hir_map.node_to_string(node_id)));
+                    self.hir_map.def_path(DefId::local(hir_id.owner)).to_string_no_crate(),
+                    hir_id.local_id.as_usize(),
+                    self.hir_map.hir_to_string(prev),
+                    self.hir_map.hir_to_string(hir_id)));
```

Comment:
And keep the `node_to_string` here.

Reply:
I don't think this check work with just a `HirId` as an input. Just remove it, but keep the insert call. We can then change `hir_ids_seen` to a `FxHashSet`. In `check`, just reconstruct the `HirId` based on the key instead of the `hir_ids_seen.values()` call.

Reply:
Check removed; I'll change `hir_ids_seen` in the next PR, alongside the removal of `NodeId` from `insert`s.